File: 763-PartitionLabels/763-PartitionLabels.java
/*

763. Partition Labels
Medium

A string S of lowercase letters is given. We want to partition this string into as many parts as possible so that each letter appears in at most one part, and return a list of integers representing the size of these parts.

Example 1:

Input: S = "ababcbacadefegdehijhklij"
Output: [9,7,8]
Explanation:
The partition is "ababcbaca", "defegde", "hijhklij".
This is a partition so that each letter appears in at most one part.
A partition like "ababcbacadefegde", "hijhklij" is incorrect, because it splits S into less parts.

Note:

    S will have length in range [1, 500].
    S will consist of lowercase letters ('a' to 'z') only.

*/

class Solution {
    public List<Integer> partitionLabels(String S) {
        // 2 passes
        // record the last position of an element in the first pass.
        // in the second pass, fill the list and return
        List<Integer> resList = new ArrayList<>();
        if(!validateInput(S)) return resList;
        int[] map = new int[26];
        for(int i = 0; i < S.length(); i++)
            map[S.charAt(i) - 'a'] = i;
        int count = 0;
        int last = 0;
        for(int i = 0; i < S.length(); i++){
            last = Math.max(last, map[S.charAt(i) - 'a']);
            count++;
            if(last == i){
                resList.add(count);
                count = 0;
            }
        }
        return resList;
    }
    private boolean validateInput(String S){
        if(S == null || S.length() == 0)
            return false;
        return true;
    }
}

File: 763-PartitionLabels/763-PartitionLabels.py
"""

763. Partition Labels

You are given a string s. We want to partition the string into as many parts as possible so that each letter appears in at most one part. For example, the string "ababcc" can be partitioned into ["abab", "cc"], but partitions such as ["aba", "bcc"] or ["ab", "ab", "cc"] are invalid.

Note that the partition is done so that after concatenating all the parts in order, the resultant string should be s.

Return a list of integers representing the size of these parts.

Example 1:

Input: s = "ababcbacadefegdehijhklij"
Output: [9,7,8]
Explanation:
The partition is "ababcbaca", "defegde", "hijhklij".
This is a partition so that each letter appears in at most one part.
A partition like "ababcbacadefegde", "hijhklij" is incorrect, because it splits s into less parts.

Example 2:

Input: s = "eccbbbbdec"
Output: [10]

Constraints:

    1 <= s.length <= 500
    s consists of lowercase English letters.

"""

class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        """
        2 passes
        in the first pass, store the last location of a character
        In the second pass, return the list
        """
        res = []
        mapp = defaultdict(int)
        for i, ch in enumerate(s): mapp[ch] = i
        count, last = 0, 0
        for i, ch in enumerate(s):
            last = max(last, mapp[ch])
            count += 1
            if last == i:
                last = 0
                res.append(count)
                count = 0
        return res
